### DISCUSSION TITLE
Background job when new record is added

### DISCUSSION THREAD

**Discussion by emilianocalzada** [pocketbase/pocketbase#5235] [Q&A] [answered]
Hello, Im using pocketbase in my new SaaS and it is a pleasure to develop using it.

I have a collection named "websites", i want to run a function (that takes a few seconds to run) in the background after a new record is added. This is my code right now:

```javascript
// When new website is created, scrape it, set website is_pending to false and add new result in separate table

onRecordAfterCreateRequest((e) => {
	const result = e.record;

        // Request to scrape website
	const res = $http.send({
		url:
			'https://XXXXX.com&url=' + result.get('url'),
		method: 'GET',
		headers: {
			'content-type': 'application/json',
			'X-KEY': ''
		},
		timeout: 120 // in seconds
	});
	const jsonResponse = res.json;
	
	// Process response, save it on another table
	...

}, 'websites');
```

I'm adding a new "websites" record via the pocketbase Admin UI, and it successfully starts scraping by calling the hook, but it shows the loading spinner UNTIL my hook functions finishes. It means that my CREATE request succeds only after the hook is finished, what can i use to make the hook run in the background and don't delay the CREATE request on "websites" collection?

**Comment by ganigeorgiev** [comment]
The JSVM doesn't allow running concurrent code as part of a request hook. If you want to run your scraping request in a non-blocking manner you basically have 2 options:

- add a [cron job](https://pocketbase.io/docs/js-jobs-scheduling/) that will query the records table for new/unprocessed records (depending on your use case you may need a helper collection or extra field to keep track of the processing state for the specific record(s))
- switch to Go and execute the request in a separate [goroutine](https://go.dev/tour/concurrency/1)

**Reply by ganigeorgiev** [reply, marked as answer]
An example cron job could look something like:

```js
// run on every 5 minutes
cronAdd("scrape", "*/5 * * * *", () => {
    // fetch the first 10 pending websites to process
    //
    // maybe change "is_pending" to "status" select field to hold more than 2 states
    // in order to indicate that the website is currently being processed
    const records = $app.dao().findRecordsByFilter(
        "websites",           // collection
        "status = 'pending'", // filter
        "created",            // sort
        10,                   // limit
        0,                    // offset
    )

    for (let record of records) {
        // (optional) refetch in case multiple cron jobs overlaps
        // to prevent processing the same record multiple times
        record = $app.dao().findRecordById(record.collection().id, record.id)
        if (record.get("status") != "pending") {
            continue // picked by another cron job
        }
    
        record.set("status", "in_progress")
        $app.dao().saveRecord(record)

        // execute your task ...

        record.set("status", "completed")
        $app.dao().saveRecord(record)
    }
})
```

_edit: added optional overlap check_

**Reply by emilianocalzada** [reply]
Thanks for your quick response! will migrate my js custom hooks and extend it pocketbase using go
